### PR TITLE
Fixes #30028 - Properly clean composed facts

### PR DESCRIPTION
**Pull request by tbrisker**
Composed `FactName`s may have associated `FactValue`s in some cases. If they
do, they shouldn't be considered orphaned and shouldn't be cleaned up.
If a composed `FactName` has no `fact_values` or only `nil` `fact_values` and no
children, it can also be considered orphaned and cleaned up.

### CODE REVIEW

**Comment by tbrisker** [comment]
Actually I think i realized why we have null fact values - they are needed so top-level composes are shown on the index (i think?)

**Comment by tbrisker** [comment]
test failures unrelated

**Comment by ezr-ondrej** [comment]
[test foreman]

**Comment by ezr-ondrej** [comment]
@domitea or @lzap would you mind taking a look?

**Comment by pmoravec** [comment]
I did few tests on two foreman / Satellite instances where both have non-leaf facts with values - all tests cleaned expected data, no uncaught exception.

**Comment by lzap** [comment]
> Actually I think i realized why we have null fact values - they are needed so top-level composes are shown on the index (i think?)

Yeah everytime I see this code I sigh. If reports is successful then I think we need to do the same with facts, instead storing them in separate tables, we would simply dump them as Postgresql hstore columns:

https://guides.rubyonrails.org/active_record_postgresql.html#hstore
https://www.postgresql.org/docs/current/hstore.html

Let me review this now.

**Comment by lzap** [comment]
I wonder what is the added value of storing facts in a tree structure. If that is for browsing experience in the UI, we can easily recreate the tree on the UI side. I think the flat structure even in the UI is fine overall. Sigh.

**Comment by ekohl** [comment]
Probably autocomplete as well.

**Comment by lzap** [comment]
Allright, I have a clean database just with few ansible facts and what I see are leaf records with and without compose flag:

```
 #<ForemanAnsible::FactName:0x0000000014bdca18
  id: 22,
  name: "ansible_all_ipv6_addresses",
  updated_at: Wed, 13 Oct 2021 17:49:13 UTC +00:00,
  created_at: Wed, 13 Oct 2021 17:49:13 UTC +00:00,
  compose: false,
  short_name: "ansible_all_ipv6_addresses",
  type: "ForemanAnsible::FactName",
  ancestry: nil>,
 #<ForemanAnsible::FactName:0x0000000014bdc838
  id: 23,
  name: "ansible_apparmor",
  updated_at: Wed, 13 Oct 2021 17:49:13 UTC +00:00,
  created_at: Wed, 13 Oct 2021 17:49:13 UTC +00:00,
  compose: true,
  short_name: "ansible_apparmor",
  type: "ForemanAnsible::FactName",
  ancestry: nil>,
```

This is weird.

**Comment by lzap** [comment]
> Probably autocomplete as well.

You can do autocomplete just fine with flat facts stored as simply `node::node::leaf` using SQL like. Anyway, the purpose of this PR is not to recreate this from scratch, but the way we store this is extremely slow. Similarly to reports, facts need to be optimized for fast inserts/updates and not for fast selects.

**Comment by lzap** [comment]
> This is weird.

Oh now I get it, `ansible_apparmor` actually is a node not leaf.

**Comment by lzap** [comment]
Okay let's setup all three cases:

```
[26] pry(main)> FactName.create(name: "DeleteMeNode", compose: true, short_name: "DeleteMeNode", type: "ForemanAnsible::FactName", ancestry: nil)
=> #<ForemanAnsible::FactName:0x0000000004a48180
 id: 1292,
 name: "DeleteMeNode",
 updated_at: Wed, 20 Oct 2021 17:49:31 UTC +00:00,
 created_at: Wed, 20 Oct 2021 17:49:31 UTC +00:00,
 compose: true,
 short_name: "DeleteMeNode",
 type: "ForemanAnsible::FactName",
 ancestry: nil>
[27] pry(main)> FactName.create(name: "DeleteMeLeaf", compose: false, short_name: "DeleteMeLeaf", type: "ForemanAnsible::FactName", ancestry: "1292")
=> #<ForemanAnsible::FactName:0x0000000015425940
 id: 1293,
 name: "DeleteMeLeaf",
 updated_at: Wed, 20 Oct 2021 17:50:04 UTC +00:00,
 created_at: Wed, 20 Oct 2021 17:50:04 UTC +00:00,
 compose: false,
 short_name: "DeleteMeLeaf",
 type: "ForemanAnsible::FactName",
 ancestry: "1292">
[28] pry(main)> FactName.create(name: "DeleteMeRootLeaf", compose: false, short_name: "DeleteMeRootLeaf", type: "ForemanAnsible::FactName", ancestry: nil)
=> #<ForemanAnsible::FactName:0x0000000015a42990
 id: 1294,
 name: "DeleteMeRootLeaf",
 updated_at: Wed, 20 Oct 2021 17:50:46 UTC +00:00,
 created_at: Wed, 20 Oct 2021 17:50:46 UTC +00:00,
 compose: false,
 short_name: "DeleteMeRootLeaf",
 type: "ForemanAnsible::FactName",
 ancestry: nil>
[29] pry(main)> FactName.where("short_name LIKE 'DeleteMe%'").all.pluck(:id)
=> [1293, 1292, 1294]
```

Now I am going to execute the rake task and see if these gets deleted.

**Comment by lzap** [comment]
Yeah it deleted them all:

```
[30] pry(main)> FactName.where("short_name LIKE 'DeleteMe%'").all.pluck(:id)
=> []
```

File: app/services/fact_cleaner.rb [hunk 1]
```diff
@@ -57,4 +74,17 @@ def delete_excluded_facts
       @deleted_count += delete_facts_names_values(fact_name_ids)
     end
   end
+
+  def valueless_fact_names
+    FactName.where.not(id: FactValue.reorder(nil).distinct.select(:fact_name_id)).reorder(nil)
```
Comment (lzap) [inline app/services/fact_cleaner.rb:79]: For the record: `SELECT "fact_names".* FROM "fact_names" WHERE "fact_names"."id" NOT IN (SELECT DISTINCT "fact_values"."fact_name_id" FROM "fact_values")`

File: app/services/fact_cleaner.rb [hunk 2]
```diff
@@ -31,23 +31,40 @@ def delete_facts_names_values(fact_name_ids)
...
+    update_leaves
+  end
+
+  def delete_leaf_orphaned_facts
     logger.debug "Cleaning leaf orphaned facts"
-    FactName.leaves.preload(:fact_values).reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
-      fact_name_ids = batch.select { |fact| fact.fact_values.empty? }
-      @deleted_count += log_removed(FactName.unscoped.where(:id => fact_name_ids).delete_all, "fact name")
+    valueless_fact_names.leaves.in_batches(of: @batch_size) do |batch|
+      @deleted_count += log_removed(batch.delete_all, "fact name")
```
Comment (lzap) [inline app/services/fact_cleaner.rb:42]: Nitpick: "fact name" could be "leaf fact name". Not gonna request a rebase tho.

File: app/services/fact_cleaner.rb [hunk 3]
```diff
@@ -31,23 +31,40 @@ def delete_facts_names_values(fact_name_ids)
...
+    delete_compose_orphaned_facts
+    update_leaves
+  end
+
+  def delete_leaf_orphaned_facts
     logger.debug "Cleaning leaf orphaned facts"
-    FactName.leaves.preload(:fact_values).reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
-      fact_name_ids = batch.select { |fact| fact.fact_values.empty? }
-      @deleted_count += log_removed(FactName.unscoped.where(:id => fact_name_ids).delete_all, "fact name")
+    valueless_fact_names.leaves.in_batches(of: @batch_size) do |batch|
```
Comment (lzap) [inline app/services/fact_cleaner.rb:41]: Renders to: `SELECT "fact_names".* FROM "fact_names" WHERE "fact_names"."id" NOT IN (SELECT DISTINCT "fact_values"."fact_name_id" FROM "fact_values") AND "fact_names"."compose" = $1  [["compose", false]]`

File: app/services/fact_cleaner.rb [hunk 4]
```diff
@@ -57,4 +74,17 @@ def delete_excluded_facts
...
+  end
+
+  def live_ancestors
+    return @live_ancestors if @live_ancestors
+    @live_ancestors = Set.new
+    FactName.leaves.reorder(nil).select(FactName.ancestry_base_class.ancestry_column, :id).find_each(:batch_size => @batch_size) do |leaf|
+      @live_ancestors.merge(leaf.ancestor_ids)
+    end
+    @live_ancestors
+  end
```
Comment (lzap) [inline app/services/fact_cleaner.rb:89]: This loads all IDs into memory, BUT we do this from a Rake task, that is okay.

File: app/services/fact_cleaner.rb [hunk 5]
```diff
@@ -31,23 +31,40 @@ def delete_facts_names_values(fact_name_ids)
...
 
+  def delete_compose_orphaned_facts
     logger.debug "Cleaning compose orphaned facts"
-    live_ancestors = Set.new
-    FactName.leaves.reorder(nil).select(FactName.ancestry_base_class.ancestry_column, :id).find_each(:batch_size => @batch_size) do |leaf|
-      live_ancestors.merge(leaf.ancestor_ids)
+
+    # Delete the composes with no values or children
+    valueless_fact_names.composes.where.not(id: live_ancestors).in_batches(of: @batch_size) do |batch|
```
Comment (lzap) [inline app/services/fact_cleaner.rb:50]: This is something like: `SELECT "fact_names".* FROM "fact_names" WHERE "fact_names"."id" NOT IN (SELECT DISTINCT "fact_values"."fact_name_id" FROM "fact_values") AND "fact_names"."compose" = $1 AND "fact_names"."id" NOT IN ($2, $3, $4, $5)  [["compose", true], ["id", 1], ["id", 2], ["id", 3], ["id", 4]]`

File: app/services/fact_cleaner.rb [hunk 6]
```diff
@@ -31,23 +31,40 @@ def delete_facts_names_values(fact_name_ids)
...
+      @deleted_count += log_removed(batch.delete_all, "fact name")
     end
-    FactName.composes.reorder(nil).where.not(id: live_ancestors).select(:id).find_in_batches(:batch_size => @batch_size) do |batch|
-      fact_name_ids = batch.map(&:id)
-      @deleted_count += log_removed(FactName.unscoped.where(:id => fact_name_ids).delete_all, "fact name")
+
+    # Delete composes that have no children and only null values
+    FactName.composes.reorder(nil).where.not(id: live_ancestors).preload(:fact_values).select(:id).find_in_batches(:batch_size => @batch_size) do |names_batch|
```
Comment (lzap) [inline app/services/fact_cleaner.rb:55]: Also good: `SELECT "fact_names"."id" FROM "fact_names" WHERE "fact_names"."compose" = $1 AND "fact_names"."id" NOT IN ($2, $3, $4, $5)  [["compose", true], ["id", 1], ["id", 2], ["id", 3], ["id", 4]]`

File: app/services/fact_cleaner.rb [hunk 7]
```diff
@@ -31,23 +31,40 @@ def delete_facts_names_values(fact_name_ids)
...
+      fact_name_ids = names_batch.select do |fact_name|
+        fact_name.fact_values.all? { |fact_value| fact_value.value.nil? }
+      end.map(&:id)
+      @deleted_count += delete_facts_names_values(fact_name_ids)
     end
   end
 
+  # If you have a value and no descendants, you're a leaf, not a compose!
```
Comment (lzap) [inline app/services/fact_cleaner.rb:63]: Sir, yes, sir! :-)

**Review by lzap — APPROVED** [review]
Allright, nice.

I mean, nice patch. Horrible code, I am sure we can do better next time :-)

**Pull request merged**